Adding Excel Copilot example

yanisoln opened the pull request:
Hey!
I know this idea wasn't listed in the original issue, but after doing a bit of research and building a prototype, I found it both fun and useful, so I thought it was worth sharing here.

This PR introduces a basic implementation of the “Excel Copilot” example I mentioned [here](https://github.com/mediar-ai/terminator/issues/24#issuecomment-2922627043), using Terminator + Tauri.

It can read a PDF, extracts structured data, and fills in the Excel UI live through the SDK toolcalls. I focused on making the interaction smooth and visible to the user to highlight the SDK’s potential in "Copilot-style" automation.

It’s probably not perfect, and I’m sure there are things that could be improved, feel free to let me know what you'd like to see changed or adjusted.
Happy to iterate or adapt it more closely to your vision!

Here is a video showing it in action :

https://github.com/user-attachments/assets/06069764-6005-443f-8dd8-851b5e185113

/claim #24

yanisoln commented:
> something i learned with excel / google sheet also is that you can paste tables formatted in TSV format so you can create a tool for the LLM to paste TSV in the spreadsheet to fill stuff quite fast

I think that's actually a good idea to let the LLM create data faster and save on tokens, will do some research and try to implement it.
As for Google Sheets, I actually tried making it work too, but unfortunately the cells aren’t exposed via UIA, so it’s really tricky to target or interact with them directly in the same way as the Excel desktop app.

> another idea i had with excel / sheets is to create a tool for the terminator LLM to actually use Microsoft Copilot or Sheets Gemini lol, maybe for some use cases

This could probably be useful for trickier things, like asking directly to copilot to use style or conditional style on the sheet, or maybe add graphs. I think it's worth exploring for some cases.

> also did not read the full code but i recommend maintaining each app/window full tree in the prompt so the LLM has full context and can use IDs as selector to interact with specific element

The good thing with Excel is that the cells have their name as automation id (A1, C5, D8, etc...) so it's really easy to target them, so unless we have to interact with the ribbon (which is not needed since Excel have shortcuts for barely anything), I think including the whole tree would mostly just burn through tokens for not much gain.

> also when merging would be cool to be able to provide a .exe / .dmg for ppl to try out this easily (i can help on this)

Not a problem, I can provide that

louis030195 commented:
i guess for google sheets its easy to paste data using TSV but to read you can use gemini that tool call the actual gemini of sheets lol

yanisoln commented:
Okay, so I added interaction with MS Copilot (for styling, conditional formatting, etc.), and google sheet support by… asking Gemini to talk to Gemini (lol). That said, Gemini in Google Sheets is quirky, it struggles with follow-ups and sometimes does random stuff or write result in the chat instead of the sheet (as seen in the second video).

Also I added TSV pasting from the LLM to speed things up, and packaged a .exe for Windows that should (normally) work anywhere, as long as webview2 is installed, let me know if that's not the case.

Finally, tried building a .dmg for macOS, but it doesn’t work yet, the macOS app will crash as soon as it try to interact with the Excel window, I will probably open another PR to add proper macOS support and push the prototype further when i'll find some time.

https://github.com/user-attachments/assets/8ea554e3-2e41-4bc6-bb8a-f6190f9171ad


https://github.com/user-attachments/assets/6ea88730-d51b-455e-97dc-ace2c1b2fafe

I'll also add some UI improvement and fix some UI related bugs.

Let me know what you think!

louis030195 commented:
some issues:

- api key keep being lost (i do a few message then it's lost)

<img width="1512" alt="Screenshot 2025-06-05 at 11 25 02 AM" src="https://github.com/user-attachments/assets/ae4fc95c-a372-4a83-a5d0-c15b787fa6d3" />
<img width="1512" alt="Screenshot 2025-06-05 at 11 26 02 AM" src="https://github.com/user-attachments/assets/bfbcd645-8b4d-4af8-97f9-6807605312e6" />
<img width="1512" alt="Screenshot 2025-06-05 at 11 26 53 AM" src="https://github.com/user-attachments/assets/d0b48116-a873-4eae-a3f1-2730687996c5" />

- cannot scroll up
- sometimes i could not see the input chat when i was in split screen (some responsive issue)
<img width="1512" alt="Screenshot 2025-06-05 at 11 28 18 AM" src="https://github.com/user-attachments/assets/bdce20ba-f368-4596-8b0e-44bd0899980f" />


could not test anything else, did not work :/

yanisoln commented:
> * api key keep being lost (i do a few message then it's lost)

Okay that's weird, i'm going to look into it, same for the scrolling up issue, also are you testing from the code directly or the binary in `excel-copilot/bin` ?

> * sometimes i could not see the input chat when i was in split screen (some responsive issue)

I think it's because the app consider you did not configure Gemini yet bc you got an error on your last message, when it's the case the input text bar does not appear, so that is linked with the previous bug
![image](https://github.com/user-attachments/assets/28506070-19a4-4361-b65b-0b8f111f138a)

Will look into it and be back with fixes.

yanisoln commented:
Update :
I improved the responsiveness of the app (especially in split-screen mode), so it should now behave better.

The bug with scrolling should be gone.

Regarding the issue with the API key being lost, I wasn’t able to reproduce it on my side. I managed to send several messages with Gemini without any problem. I’m not exactly sure what caused the issue, it might be related to how the API key was obtained or configured in your environment.

That said, I added a second provider as a fallback: you can now use OpenAI’s API (GPT-4.1 mini) if needed. It should offer a more stable alternative in case Gemini runs into issues.

Let me know how it goes if you test it again! 🙌

Anshgrover23 commented:
@yanisoln can u add a new video, by using the property highlight element as it is becoming difficult to identify the difference b/w your hand drawn changes and the terminator changes(automatic cursor changes or clicks) lol, happy to review more after that.

yanisoln commented:
> @yanisoln can u add a new video, by using the highlight method as it is becoming difficult to identify the difference b/w your hand drawn changes and the terminator changes(automatic cursor changes or clicks) lol, happy to review more after that.

Hey! Thanks for the feedback. But i'm not 100% sure I understand what you're asking. Just to clarify, I didn't touch my mouse or keyboard at all during the videos while the app was working (except between prompts, of course). Everything was triggered by the LLM with toolcalls powered by the SDK.

Anshgrover23 commented:
@yanisoln try running this example, to know how highlight works.  https://github.com/mediar-ai/terminator/blob/main/examples/gmail_automation.py

yanisoln commented on examples/excel-copilot/package.json:
```diff
@@ -0,0 +1,31 @@
+    "tauri:build": "tauri build"
+  },
+  "dependencies": {
+    "@tauri-apps/api": "^2.5.0",
```
So far everything's been working fine without `@types/node`, and I’m not using any Node APIs on frontend, so i don't see any particular reason to include it.
Let me know if I missed smth!

yanisoln replied:
Oh okay, I didn’t have any issue on my side, but if it solves the problem on yours, I don’t mind adding it, doesn’t cost much anyway.
